###################################################################################################
##  A complete R-wrap for identification of differentially used exons and exon-exon junctions    ##
##  using functionality of the JunctionSeq package.                                              ##
##  (c) GNU GPL Vasily V. Grinev, 2018-2019. grinev_vv[at]bsu.by                                 ##
###################################################################################################
##  Setting of work directory.
setwd("D:/Software/QoRTs")
d.work = "D:/Software/QoRTs"
##  Loading of required auxiliary library.
#   This code was successfully tested with library JunctionSeq v.1.14.0.
suppressMessages(require(JunctionSeq))
##  Generating raw counts via QoRTs v.1.3.6 (via CMD shell).
java -Xmx7G -jar D:/Software/QoRTs/QoRTs.jar QC \
                 --singleEnded \
                 --maxReadLength 76 \
                 --minMAPQ 30 \
                 --verbose \
                 --noGzipOutput \
                 --runFunctions writeKnownSplices,writeNovelSplices,writeSpliceExon \
                 D:/Software/QoRTs/Files_BAM/sample1.bam \
                 D:/Software/QoRTs/Files_GTF/Ensembl_GRCh38.p7_release.85.gtf \
                 D:/Software/QoRTs/Files_COUNTS/sample1/
##  Generation a special flat GFF file and a new set of JunctionSeq-formatted count files
#   containing all exons, annotated splice junctions
#   and passed-filter novel splice junctions (via CMD shell).
java -Xmx7G -jar D:/Software/QoRTs/QoRTs.jar mergeNovelSplices \
                 --minCount 10 \
                 --minSpan 50 \
                 --noGzipInput \
                 --noGzipOutput \
                 --verbose \
                 D:/Software/QoRTs/Files_COUNTS/ \
                 D:/Software/QoRTs/Files_COUNTS/decoder.txt \
                 D:/Software/QoRTs/Files_GTF/Ensembl_GRCh38.p7_release.85.gtf \
                 D:/Software/QoRTs/Files_COUNTS/
##  Loading of annotation file.
anno = read.table(file = paste(d.work, "Files_COUNTS/decoder.txt", sep = "/"),
                  sep = "\t",
                  header = TRUE,
                  quote = "\"",
                  as.is = TRUE)
##  Listing of the count files generated by QoRTs.
f.counts = paste0(paste(d.work, "Files_COUNTS/", sep = "/"),
                  anno$sample.ID,
                  "/QC.spliceJunctionAndExonCounts.withNovel.forJunctionSeq.txt")
##  A flattened gff-formatted annotation file from which the gene counts were generated.
f.gff = paste(d.work, "Files_COUNTS/withNovel.forJunctionSeq.gff", sep = "/")
##  Run a complete JunctionSeq analysis from start to finish.
jscs = runJunctionSeqAnalyses(sample.files = f.counts,
                              sample.names = anno$sample.ID,
                              condition = factor(anno$group.ID),
                              flat.gff.file = f.gff,
                              analysis.type = "junctionsAndExons",
                              verbose = TRUE)
##  Generation of output data files.
writeCompleteResults(jscs = jscs,
                     outfile.prefix = "KD_",
                     gzip.output = FALSE,
                     FDR.threshold = 0.05,
                     save.allGenes = TRUE,
                     save.sigGenes = TRUE,
                     save.fit = TRUE,
                     save.VST = TRUE,
                     save.bedTracks = TRUE,
                     save.jscs = TRUE,
                     bedtrack.format = "GTF",
                     verbose = TRUE)
color = list(SIG.FEATURE.COLOR = rgb(213, 94, 0, maxColorValue = 255),
             SIG.VERTLINE.COLOR = rgb(213, 94, 0, maxColorValue = 255),
             SIG.FEATURE.BORDER.COLOR = "black",
             SIG.FEATURE.FILL.COLOR = rgb(213, 94, 0, maxColorValue = 255),
             NOSIG.FEATURE.COLOR = "black",
             NOSIG.VERTLINE.COLOR = "black",
             NOSIG.FEATURE.BORDER.COLOR = "black",
             NOSIG.FEATURE.FILL.COLOR = rgb(0, 158, 115, maxColorValue = 255),
             UNTESTABLE.FEATURE.COLOR = rgb(0, 114, 178, maxColorValue = 255),
             UNTESTABLE.VERTLINE.COLOR = rgb(0, 114, 178, maxColorValue = 255),
             UNTESTABLE.FEATURE.BORDER.COLOR = "black",
             UNTESTABLE.FEATURE.FILL.COLOR = rgb(0, 114, 178, maxColorValue = 255))
##  Generation and saving of JunctionSeq expression plots.
#   Loading a list of target genes (if necessary).
t.genes = read.table(file = paste(d.work, "List of target genes.txt", sep = "/"),
                     sep = "\t",
                     header = TRUE,
                     quote = "\"",
                     as.is = TRUE)$gene_id
#   Building plots.
buildAllPlots(jscs = jscs,
              outfile.prefix = paste(d.work, "Plots", sep = "/"),
#              FDR.threshold = 0.05,
              gene.list = t.genes,
#              max.gene.ct = 20,
              use.plotting.device = "png",
              sequencing.type = "single-end",
              use.log = TRUE,
              exon.rescale.factor = 0.3,
              subdirectories.by.type = TRUE,
              ma.plot = FALSE,
              variance.plot = FALSE,
              with.TX = TRUE,
              without.TX = FALSE,
              expr.plot = TRUE,
              normCounts.plot = FALSE,
              rExpr.plot = FALSE,
              rawCounts.plot = FALSE,
              colorList = color,
              plot.gene.level.expression = TRUE,
              plot.lwd = 3,
              axes.lwd = 3,
              anno.lwd = 3,
              gene.lwd = 3,
              par.cex = 1,
              anno.cex.text = 1,
              anno.cex.axis = 1,
              anno.cex.main = 1.2,
              drawCoordinates = TRUE,
              yAxisLabels.inExponentialForm = FALSE,
              show.strand.arrows = 1,
              GENE.annotation.relative.height = 0.15,
              name.files.with.geneID = TRUE,
              minimalImageFilenames = FALSE,
              verbose = TRUE)
##  Generating a MA-plot.
plotMA(jscs = jscs,
       FDR.threshold = 0.05,
       fc.name = NULL,
       fc.thresh = 1,
       use.pch = 19,
       smooth.nbin = 256,
       use.smoothScatter = TRUE,
       ylim = c(1/100, 100),
       label.counts = TRUE,
       label.axes = c(TRUE, TRUE, FALSE, FALSE),
       show.labels = TRUE,
       par.cex = 1,
       points.cex = 1.7,
       text.cex = 1,
       lines.cex = 8,
       anno.lwd = 0.1,
       mar = c(4.1, 4.1, 3.1, 1.1),
       miniTicks = TRUE,
       verbose = TRUE)
